### PR TITLE
CB-14585 ccm upgrade revert flow in case of error in freeipa

### PR DESCRIPTION
**Pull request by afarsang**

### CODE REVIEW

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java [hunk 1]
```diff
@@ -214,16 +222,39 @@ private void verifyServiceCalls(int calledOnceCount) throws CloudbreakOrchestrat
         inOrder.verify(upgradeCcmService, times(expected[i++])).changeTunnel(STACK_ID);
         inOrder.verify(upgradeCcmService, times(expected[i++])).obtainAgentDataState(STACK_ID);
         inOrder.verify(upgradeCcmService, times(expected[i++])).obtainAgentData(STACK_ID);
+        if (calledOnceCount == CALLED_ONCE_TILL_OBTAIN_AGENT_DATA) {
```
Comment (bergerdenes) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:225]: Can you please guide me through this?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/ccm/CcmParametersConfigService.java [hunk 2]
```diff
@@ -23,8 +24,13 @@ public class CcmParametersConfigService {
 
     private static final String UPGRADE_CCM_KEY = "ccm_jumpgate";
 
+    private static final String UPGRADE_CCM_REVERT_JOB_KEY = "revert_job";
```
Comment (bergerdenes) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/ccm/CcmParametersConfigService.java:27]: is it used?

**Review by bergerdenes — CHANGES_REQUESTED** [review]
Please have a discussion about explaining the complete workflow.
It would be worth documenting it, too.

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/AbstractUpgradeCcmAction.java [hunk 3]
```diff
@@ -51,16 +52,18 @@ protected UpgradeCcmContext createFlowContext(FlowParameters flowParameters, Sta
     @Override
     protected Object getFailurePayload(P payload, Optional<UpgradeCcmContext> flowContext, Exception e) {
         Tunnel oldTunnel = null;
+        LocalDateTime reverTime = null;
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/AbstractUpgradeCcmAction.java:55]: ok

File: freeipa/src/main/resources/freeipa-salt/salt/upgradeccm/restoressl.sls [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    - failhard: True
+    - watch:
+      - file: /etc/nginx/sites-enabled/ssl.conf
+
+{#
+nginx_restart:
+  cmd.run:
+    - name: "systemctl restart nginx"
+    - failhard: True
+#}
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/resources/freeipa-salt/salt/upgradeccm/restoressl.sls:25]: ok

File: freeipa/src/main/resources/freeipa-salt/salt/upgradeccm/restoressl.sls [hunk 5]
```diff
@@ -0,0 +1,25 @@
+restore_last_ssh_conf:
+  module.run:
+    - name: file.restore_backup
+    - path: /etc/nginx/sites-enabled/ssl.conf
+    - backup_id: 0
+
+touch_ssl_conf_for_restart:
+  file.touch:
+    --name: /etc/nginx/sites-enabled/ssl.conf
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/resources/freeipa-salt/salt/upgradeccm/restoressl.sls:9]: this is wrong

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/ccm/UpgradeCcmOrchestratorService.java [hunk 6]
```diff
@@ -55,6 +57,12 @@ public void reconfigureNginx(Long stackId) throws CloudbreakOrchestratorExceptio
         hostOrchestrator.runOrchestratorState(stateParams);
     }
 
+    public void finalizeConfiguration(Long stackId) throws CloudbreakOrchestratorException {
+        OrchestratorStateParams stateParams = createStateParams(stackId, FINALIZE);
+        LOGGER.debug("Calling reconfigureNginx with state params '{}'", stateParams);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/ccm/UpgradeCcmOrchestratorService.java:62]: ok

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java [hunk 7]
```diff
@@ -105,6 +104,10 @@ public void changeTunnel(Long stackId) {
         stackService.setTunnelByStackId(stackId, Tunnel.latestUpgradeTarget());
     }
 
+    public void changeTunnel(Long stackId, Tunnel tunnel) {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java:107]: I've  chosen the later

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java [hunk 8]
```diff
@@ -126,7 +129,6 @@ public void pushSaltStatesState(Long stackId) {
     }
 
     public void pushSaltStates(Long stackId) throws CloudbreakOrchestratorException {
-        InMemoryStateStore.putStack(stackId, PollGroup.POLLABLE);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java:129]: I think it was a mistake only

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java [hunk 9]
```diff
@@ -181,17 +186,17 @@ protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map
     public Action<?, ?> finished() {
         return new AbstractUpgradeCcmEventAction() {
             @Override
-            protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map<Object, Object> variables) {
+            protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map<Object, Object> variables) throws CloudbreakOrchestratorException {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java:189]: yeah, this was a sloppy solution, I rewrite this

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java [hunk 10]
```diff
@@ -53,12 +63,38 @@ public Class<? extends AbstractUpgradeCcmEventHandler> getFailureOrigin() {
...
+        return this;
+    }
+
     @Override
     public String toString() {
         return "UpgradeCcmFailureEvent{" +
                 "oldTunnel=" + oldTunnel +
                 ", transitionStatusAfterFailure=" + transitionStatusAfterFailure +
                 ", failureOrigin=" + failureOrigin +
+                ", statusReason=" + statusReason +
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java:93]: we do

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java [hunk 11]
```diff
@@ -13,17 +14,23 @@ public class UpgradeCcmFailureEvent extends StackFailureEvent {
 
     private final Tunnel oldTunnel;
 
+    private LocalDateTime revertTime;
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java:17]: final....

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java [hunk 12]
```diff
@@ -53,12 +63,38 @@ public Class<? extends AbstractUpgradeCcmEventHandler> getFailureOrigin() {
...
+
+    public void setStatusReason(String statusReason) {
+        this.statusReason = statusReason;
+    }
+
+    public LocalDateTime getRevertTime() {
+        return revertTime;
+    }
+
+    public UpgradeCcmFailureEvent withStatusReason(String statusReason) {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java:82]: it was nice to use this ' EVENT.createBasedOn(previousEvent).withStatusReason(""). But I can remove it, is you don't like it

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmDeregisterMinaHandler.java [hunk 13]
```diff
@@ -34,15 +34,21 @@ public String selector() {
...
     @Override
     protected Selectable doAccept(HandlerEvent<UpgradeCcmEvent> event) {
         UpgradeCcmEvent request = event.getData();
         if (request.getOldTunnel().useCcmV1()) {
             LOGGER.info("Deregistering mina for CCM upgrade...");
-            upgradeCcmService.deregisterMina(request.getResourceId());
+            try {
+                upgradeCcmService.deregisterMina(request.getResourceId());
+            } catch (Exception e) {
+                LOGGER.info("Deregistering mina is failed, but stack is operation in CCMv2");
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmDeregisterMinaHandler.java:49]: OK

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmRemoveMinaHandler.java [hunk 14]
```diff
@@ -46,10 +46,12 @@ protected Selectable doAccept(HandlerEvent<UpgradeCcmEvent> event) {
...
                 upgradeCcmService.removeMina(request.getResourceId());
-            } catch (CloudbreakOrchestratorException e) {
-                LOGGER.debug("Failed removing Mina service with a salt state");
-                return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_EVENT.event(), request.getResourceId(),
-                        event.getData().getOldTunnel(), getClass(), e);
+            } catch (Exception e) {
+                LOGGER.debug("Removing Mina for CCM upgrade has been failed, but new CCM is operational.");
+//                return new UpgradeCcmFailureEvent(UPGRADE_CCM_CLEANING_FAILED_EVENT.event(), request.getResourceId(),
+//                        event.getData().getOldTunnel(), getClass(), e, event.getData().getRevertTime(),
+//                        "Removing Mina for CCM upgrade has been failed, but new CCM is operational.");
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmRemoveMinaHandler.java:53]: right, removed

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmRemoveMinaHandler.java [hunk 15]
```diff
@@ -46,10 +46,12 @@ protected Selectable doAccept(HandlerEvent<UpgradeCcmEvent> event) {
             LOGGER.info("Remove Mina for CCM upgrade...");
             try {
                 upgradeCcmService.removeMina(request.getResourceId());
-            } catch (CloudbreakOrchestratorException e) {
-                LOGGER.debug("Failed removing Mina service with a salt state");
-                return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_EVENT.event(), request.getResourceId(),
-                        event.getData().getOldTunnel(), getClass(), e);
+            } catch (Exception e) {
+                LOGGER.debug("Removing Mina for CCM upgrade has been failed, but new CCM is operational.");
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmRemoveMinaHandler.java:50]: ok

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmRemoveMinaHandler.java [hunk 16]
```diff
@@ -35,8 +34,9 @@ public String selector() {
     @Override
     protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpgradeCcmEvent> event) {
         LOGGER.error("Removing Mina for CCM upgrade has failed", e);
-        return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_EVENT.event(), resourceId,
-                event.getData().getOldTunnel(), getClass(), e);
+        return new UpgradeCcmFailureEvent(UPGRADE_CCM_CLEANING_FAILED_EVENT.event(), resourceId,
+                event.getData().getOldTunnel(), getClass(), e, event.getData().getRevertTime(),
+                "Removing Mina for CCM upgrade has been failed, but new CCM is operational.");
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmRemoveMinaHandler.java:39]: ok

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java [hunk 17]
```diff
@@ -161,33 +165,37 @@ public void testCcmUpgradeWhenRegisterCcmFails() throws CloudbreakOrchestratorEx
 
     @Test
     public void testCcmUpgradeWhenHealthCheckFails() throws CloudbreakOrchestratorException {
-        doThrow(new BadRequestException()).when(upgradeCcmService).healthCheck(1L);
+        doThrow(new BadRequestException()).doNothing().when(upgradeCcmService).healthCheck(1L);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:168]: Yes, there is two health check call in this test:
1)healthcheck after upgrade (doThrow Exception)
2)healthcheck after revert (doNothing => old ccm is working again)

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java (hunk 1 above)
Comment (afarsang) [reply freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:225]: Yes, It is not the most elegant,but these are the revert cases for every cases. so the test will pass calledOnceCount meaning this is where exception will happen, and (after this all the call count will be 0) checking the calls for the revert branch.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/ccm/CcmParametersConfigService.java (hunk 2 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/upgrade/ccm/CcmParametersConfigService.java:27]: no,removed

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (bergerdenes) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmFailureEvent.java:74]: I don't like it :)

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java [hunk 18]
```diff
@@ -183,15 +191,52 @@ protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map
             @Override
             protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map<Object, Object> variables) {
                 LOGGER.info("FreeIPA CCM upgrade finished {}", payload);
-                upgradeCcmService.finishedState(context.getStack().getId());
+                try {
+                    upgradeCcmService.finishedState(context.getStack().getId(), payload.isMinaRemoved());
+                } catch (CloudbreakOrchestratorException e) {
+                    LOGGER.warn("Finalizing CCM upgrade is failed with {}", e);
+                    Class<? extends AbstractUpgradeCcmEventHandler> x = null;
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java:198]: right

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java [hunk 19]
```diff
@@ -211,23 +241,47 @@ private void verifyServiceCalls(int calledOnceCount) throws CloudbreakOrchestrat
...
+            inOrder.verify(upgradeCcmService, times(1)).registerClusterProxy(eq(STACK_ID));
+            inOrder.verify(upgradeCcmService, times(1)).healthCheck(eq(STACK_ID));
+            inOrder.verify(upgradeCcmService, times(1)).pushSaltStates(eq(STACK_ID));
+
+        }
         inOrder.verify(upgradeCcmService, times(expected[i++])).removeMinaState(STACK_ID);
         inOrder.verify(upgradeCcmService, times(expected[i++])).removeMina(STACK_ID);
         inOrder.verify(upgradeCcmService, times(expected[i++])).deregisterMinaState(STACK_ID);
         inOrder.verify(upgradeCcmService, times(expected[i++])).deregisterMina(STACK_ID);
+        //inOrder.verify(upgradeCcmService, times(expected[i++])).finishedState(any(), any());
```
[the remark this answers is not in the document]
Comment (afarsang) [reply freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:284]: no, I removed that

**Pull request closed without merging**